### ISSUE TITLE
Support PutIfNotExist commit handler for R2 and Minio

### ISSUE THREAD

**Issue by wjones127** [lancedb/lance#2246]
While S3 infamously doesn't support any atomic put_if_not_exists or similar, we have to have some supplemental commit mechanism. However, there are S3-compatible stores that do support such an operation: R2 and Minio. It would be cool to support concurrent writes out-of-the-box for these backends.

We should be able to detect whether a user is pointing at one of those services (or just let them tell us). Then we can switch to this other commit handler.

See: https://docs.rs/object_store/0.9.1/object_store/aws/enum.S3ConditionalPut.html

**Comment by rakeshJn** [comment]
Looks like IBM Cloud Object Store also supports conditional writes: https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-upload#upload-conditional
It would be nice to have this feature available, without which it is quite limited.

**Comment by metadaddy** [comment]
Amazon S3 recently (8-20-24, to be exact) introduced support for conditional writes - `PutObject` and `CompleteMultipartUpload` now accept an `If-None-Match` header. See https://docs.aws.amazon.com/AmazonS3/latest/userguide/conditional-writes.html for details.

**Comment by wjones127** [comment]
> Amazon S3 recently (8-20-24, to be exact) introduced support for conditional writes - `PutObject` and `CompleteMultipartUpload` now accept an `If-None-Match` header. See https://docs.aws.amazon.com/AmazonS3/latest/userguide/conditional-writes.html for details.

Indeed. We are tracking support for that in https://github.com/lancedb/lance/issues/2793

**Comment by wjones127** [comment]
Closed by https://github.com/lancedb/lance/pull/3483
